crates/flowctl/src/source/unbundle.rs

```diff
@@ -0,0 +1,338 @@
+        // This branch also handles the case of `Existing::Error` if the file already exists.
+        Err(err) => Err(anyhow::anyhow!("unable to open file: {}", err)),
+    }
+}
+
+async fn write_file(path: &Path, contents: &[u8], existing: Existing) -> anyhow::Result<()> {
+    match try_write_file(path, contents, existing).await {
+        Ok(()) => {
+            tracing::info!(path = %path.display(), "wrote file");
+            Ok(())
+        }
+        Err(err) => {
+            tracing::error!(path = %path.display(), error = %err, "failed to write file");
+            Err(err).context(format!("failed to write file: '{}'", path.display()))
+        }
+    }
+}
+
+async fn parse_catalog_spec(flow_yaml_path: &Path) -> Result<models::Catalog, anyhow::Error> {
+    let mut existing = fs::File::open(flow_yaml_path).await?;
+    let mut contents = Vec::with_capacity(16 * 1024);
+    existing.read_to_end(&mut contents).await?;
+    let cat = sources::parse_catalog_spec(&contents)?;
+    Ok(cat)
+}
+
+/// Returns the last path component for a given catalog name, which must be a valid catalog name.
+fn base_name(catalog_name: &impl AsRef<str>) -> &str {
+    catalog_name
+        .as_ref()
+        .rsplit_once("/")
+        .expect("catalog names have at least one '/'")
+        .1
+}
+
+async fn indirect_endpoint_config(
+    config: &mut models::ConnectorConfig,
+    dir: &Path,
+    filename: String,
+    serializer: &Serializer,
+    existing: Existing,
+) -> anyhow::Result<()> {
+    let file_path = dir.join(&filename);
+    let contents = serializer.to_vec(&config.config)?;
+
+    write_file(&file_path, &contents, existing).await?;
+    config.config = RawValue::from_string(Value::String(filename).to_string()).unwrap();
+    Ok(())
+}
+
+async fn maybe_indirect_schema(
+    dir: &Path,
+    filename: &str,
+    serializer: &Serializer,
+    schema: &mut models::Schema,
+    existing: Existing,
+) -> anyhow::Result<()> {
+    // Attempt to clean up the schema by removing a superfluous $id.
+    match schema {
+        models::Schema::Object(m) => {
+            if m.contains_key("definitions") || m.contains_key("$defs") {
+                // We can't touch $id, as it provides the canonical base against which
+                // $ref is resolved to definitions.
+            } else if let Some(true) = m
+                .get("$id")
+                .and_then(Value::as_str)
+                .map(|s| s.starts_with("file://"))
+            {
+                m.remove("$id");
+            }
+        }
+        _ => anyhow::bail!("expected object schema but got {schema:?}"),
+    };
+
+    let contents = serializer.to_vec(&schema)?;
+
+    if contents.len() <= MAX_INLINE_SIZE {
+        // Leave small schema in-place.
+        return Ok(());
+    }
+
+    let path = dir.join(filename);
+    write_file(&path, &contents, existing).await?;
+    *schema = models::Schema::Url(models::RelativeUrl::new(filename));
+
+    Ok(())
+}
+
+fn merge_catalog(
```

Comment:
You say
> The merge operation is very basic. It does not try to merge properties of individual specs, but only the overall set of specs in each file.

in the PR description, which matches what's going on here, but is this really sufficient? I'm certainly not thinking of all the ways this could be used, but what about the following scenario:

* You have a spec locally, and you have defined the capture `acmeCo/pizza/my_capture`. This capture exists already, but you're making a change to its config, maybe endpoint config or schema or something
* You realize that you actually need a bunch of other `acmeCo/pizza` specs, so you pull them down with `flowctl catalog pull-specs --prefix acmeCo/pizza/ --existing=merge-spec`.

Does this logic mean that after running `pull-specs`, your local changes to `acmeCo/pizza/my_capture` will be blown away and replaced with the specs as they exist in production?

Reply:
> Does this logic mean that after running pull-specs, your local changes to acmeCo/pizza/my_capture will be blown away and replaced with the specs as they exist in production?

Yes it does. I'm not really sure whether it's sufficient, but I also don't really see a straight forward solution that's better. We can't merge endpoint configs, and need to just pick either the local copy or the remote one. We could do something like `--existing=merge-specs-but-keep-endpoint-configs`, but it feels to me like that introduces more confusion than it's worth. Now that I'm thinking about it, probably the best default (for interactive shells, at least) would be `--existing=prompt`, which would prompt the user for each existing file. I'll propose that we implement the prompt in the next pass of UX improvements, if that makes sense to you.

Reply:
Okay cool, this makes sense. Yeah, prompting sounds like it'd be better UX, though even a simpler `git`-style failure of "You have pending changes that would be overwritten by this action. Use --force to acknowledge" would probably be good enough too